### PR TITLE
Add `no-nested-ternary` rule

### LINKED ISSUE #50: A better `no-nested-ternary` rule
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/50)

Since it was denied in ESLint: https://github.com/eslint/eslint/issues/3480

I think it would be better with a new rule that only errors on nested ternaries that are not wrapped in parens. For example: `parens-wrapped-nested-ternary` (a mouthful, but very descriptive).
## 

Thoughts?

> There is a $82.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/50).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by karaggeorge**
Fixes #50

Extends the existing eslint rule but allows one-level nesting if it's wrapped in parens

Notes:
- Wasn't sure what the best name should be.
- First time writing an eslint rule, so let me know if it needs edits
- Tested it against most of the examples in the issue and it seemed to work

Not sure about the case where both sides are ternary. Should it report two errors, one for each? Current rule only reports one for the whole expression, but if it's one-level nested, we could provide a fix for both by wrapping it in parens (current implementation)

### CODE REVIEW

File: docs/rules/no-nested-ternary.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+var foo = i > 5 ? i < 100 ? true : false : true;
+var foo = i > 5 ? true : (i < 100 ? true : (i < 1000 ? true : false));
+```
+
+
+## Pass
+
+```js
+var foo = i > 5 ? (i < 100 ? true : false) : true;
+var foo = i > 5 ? (i < 100 ? true : false) : (i < 100 ? true : false);
```
Comment (sindresorhus) [inline docs/rules/no-nested-ternary.md]: Use `const` for these.

File: docs/rules/no-nested-ternary.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Disallow nested ternary expressions
+
+Nesting ternary expressions can make code more difficult to understand.
+
+One level of nesting is allowed as long as the nested ternary is wrapped in parens
+
```
Comment (sindresorhus) [inline docs/rules/no-nested-ternary.md]: You need to mention how it's an improved version of the built-in ESLint rule (and link to it) and what it improves.

Also, at the bottom, recommend disabling the built-in rule and show how to do that.

Comment (sindresorhus) [reply docs/rules/no-nested-ternary.md]: In the readm, you marked it as "partly fixable", so would be good to explain what is and what is not fixable.

File: test/no-nested-ternary.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+const errors = [{ruleId: 'my-rule'}];
+
+ruleTester.run('new-error', rule, {
+	valid: [
+		'var foo = i > 5 ? true : false;',
+		'var foo = i > 5 ? true : (i < 100 ? true : false);',
+		'var foo = i > 5 ? (i < 100 ? true : false) : true;',
+		'var foo = i > 5 ? (i < 100 ? true : false) : (i < 100 ? true : false);',
+		'var foo = i > 5 ? true : (i < 100 ? FOO(i > 50 ? false : true) : false);'
```
Comment (sindresorhus) [inline test/no-nested-ternary.js]: Use `const`.

File: test/no-nested-ternary.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+			code: 'var foo = i > 5 ? i < 100 ? true : false : i < 100 ? true : false;',
+			output: 'var foo = i > 5 ? (i < 100 ? true : false) : (i < 100 ? true : false);',
+			errors: [{ruleId: 'my-rule', column: 19}, {ruleId: 'my-rule', column: 44}]
+		},
+		{
+			code: 'var foo = i > 5 ? true : i < 100 ? true : false;',
+			output: 'var foo = i > 5 ? true : (i < 100 ? true : false);',
+			errors
+		}
+	]
```
Comment (sindresorhus) [inline test/no-nested-ternary.js]: Would be good to take a look at the built-in ESLint rule and its test and see if there are cases that we don't handle. tl;dr Add more tests.

File: rules/no-nested-ternary.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+module.exports = {
+	create,
+	meta: {
+		type: 'suggestion',
+		docs: {
+			url: getDocsUrl(__filename)
+		},
+		fixable: 'code',
+		schema: []
```
Comment (sindresorhus) [inline rules/no-nested-ternary.js]: No point in having `scheme` here when it's empty.

File: test/no-nested-ternary.js (hunk 4 above)
Comment (karaggeorge) [reply test/no-nested-ternary.js]: They only had two of each. I added the ones that were still valid in the list

File: docs/rules/no-nested-ternary.md (hunk 2 above)
Comment (karaggeorge) [reply docs/rules/no-nested-ternary.md]: I tried to reword the readme. Let me know if that makes more sense

**Comment by karaggeorge** [comment]
@sindresorhus this should be good for another look

File: test/no-nested-ternary.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import avaRuleTester from 'eslint-ava-rule-tester';
+import rule from '../rules/no-nested-ternary';
+
+const ruleTester = avaRuleTester(test, {
+	env: {
+		es6: true
+	}
+});
+
+const errors = [{ruleId: 'my-rule'}];
```
Comment (sindresorhus) [inline test/no-nested-ternary.js]: Huh? What is `my-rule`?

---

@MrHen Seems like the `ruleId` property is not correctly checked in the tests...

Comment (sindresorhus) [reply test/no-nested-ternary.js]: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/368

**Comment by sindresorhus** [comment]
Nice work, @karaggeorge 👌

**Pull request merged**